### ISSUE TITLE
[Proposal] Catch StackOverflow with special attribute 

### ISSUE THREAD

**Issue by MkazemAkhgary** [dotnet/coreclr#14675]
Originally posted at : https://github.com/dotnet/roslyn/issues/22829#issuecomment-339082122

>Starting with the .NET Framework  version 2.0, a StackOverflowException  object cannot be caught by a try-catch  block and the corresponding process is  terminated by default. Consequently,  users are advised to write their code  to detect and prevent a stack  overflow.

**Proposal**

Why make it hard for users to deal with SO?

Some methods are prone to SO exception, not by infinite recursive, but simply because recursion becomes too deep, some times `stackalloc` can cause this exception too.

>In prior versions of the .NET  Framework, your application could  catch a StackOverflowException object  (for example, to recover from  unbounded recursion). However, that  practice is currently discouraged  because significant additional code is  required to reliably catch a stack  overflow exception and continue  program execution.

attaching special attribute should make this possible and user is aware of performance or memory overheads.

```
[CatchStackOverflowException]
void Method_SO_Catchable()
{
}
```

**Comment by jkotas** [comment]
@MkazemAkhgary How would you propose to implement this? The hard part about stack overflow handling is to ensure that there is enough stack left to run the handler.

**Comment by bendono** [comment]
> Some methods are prone to SO exception, not by infinite recursive, but simply because recursion becomes too deep, some times stackalloc can cause this exception too.

Are you able to handle these cases? Other than logging, I can't think of any reasonable way to handle such cases, which indicates crashing is preferable. A concrete example would be useful for discussion.

**Comment by MkazemAkhgary** [comment]
[Here is the actual problem that I had](https://github.com/dotnet/csharplang/issues/1040#issuecomment-339159749). and it led me to thing about this issue too. ([and this one too.](https://github.com/dotnet/roslyn/issues/22825#issuecomment-339172919))

**Comment by MkazemAkhgary** [comment]
@jkotas if stack size is 1MB, then you will have `N` extra space preserved. I cant say how much `N` would be, that's out of my knowledge. cant this size be dynamic after all? I guess in lower level not everything is same as I think so.

Another proposal is to pop every thing from top of stack (and all information will be lost, only user knows that exception happened somewhere.), a little information is kept on stack (or some where) indicating this method's SOE's can be caught. As you can see method is marked with attribute. lets call this a flag inside stack that will be pushed in every further calls by marked method. when that method is returned to caller, the flag is also popped out. 

a mix of these to proposals can be a solution too. Compiler will check if there is any method marked with such attribute, then executable stack size will have extra `N` space. if stack size is `1MB`, reaching `1MB` is guaranteed SO exception (just like before), but now there is extra space to put some information in stack trace. then Pop and catch.

@jkotas I hope that makes sense. I don't have perspective at low level programming, though ive written in assembly language once, it was scary. I don't get surprised if all these proposals are blocked by big wall :)

**Comment by poizan42** [comment]
@jkotas What are the actual issues encountered with the previous support for catching them? I'd imagine not having enough stack to run the (managed) handler means necessary cleanup would not be run - so just kill the process then. Was there complications in the native part, e.g. with avoiding the managed heap and MethodDesc's etc. to end in a corrupted stated?

**Comment by jkotas** [comment]
> @jkotas What were the actual issues encountered with the previous support for catching them?

It was very unreliable.

**Comment by jkotas** [comment]
`RuntimeHelpers.EnsureSufficientExecutionStack` is the reliable solution for handling stackoverflow that we have come up with at the end. Basically, methods that want to handle stack oveflow should to this method. Actually, it looks very similar to the attribute proposed here - just a method call vs. attribute cosmetic difference.

**Comment by jkotas** [comment]
Duplicate of https://github.com/dotnet/coreclr/issues/652

**Comment by pongba** [comment]
@jkotas this is great to know. One question: what's the performance cost?
